### ISSUE TITLE
Alias samplers (WarpLDA+LightLDA): shared Alias::build corrupts the proposal PMF for equal weights (incl. symmetric alpha) -> invalid MH; WarpLDA beta-update pending-proposal & seed-normalizer mismatches

### ISSUE THREAD

**Issue by nealcaren** [nealcaren/topica#430]
Independent correctness review of the alias-method MH samplers WarpLDA (`src/warplda.rs`, Chen et al 2016) and LightLDA (`src/lightlda.rs`, Yuan et al 2015), via the Codex second-opinion agent (`gpt-5.6-sol`). The shared `Alias` (defined in `lightlda.rs`, imported by `warplda.rs:47`) has a **verified build bug that silently corrupts the proposal distribution for equal-weight inputs — including the default symmetric prior**. The MH acceptance algebra in both samplers I re-derived by hand and found internally consistent with the papers' fixed-proposal semantics (see "Investigate" below).

## 1. [High] `Alias::build` discards a bucket when the `large` worklist empties early -> corrupted PMF

`src/lightlda.rs:74`:
```rust
while let (Some(s), Some(&l)) = (small.pop(), large.last()) { ... }
```
`small.pop()` is evaluated (and MUTATES `small`) even when `large.last()` is `None`. When `large` empties while `small` is non-empty, the final `small.pop()` removes an index that is then **never finalized**: it is gone from `small` (so the `for &s in &small { prob[s]=1.0 }` cleanup skips it), leaving `prob[s]=0.0`, `alias[s]=0`.

**Reachable with equal weights** — e.g. `[0.1, 0.1, 0.1]`: `sum = 0.30000000000000004`, `scale = 9.999...`, so every `scaled[i] = 0.999... < 1.0`. All three enter `small`, `large` is empty, the last pop is discarded, and the table represents ~`[2/3, 1/3, 0]` instead of uniform. This is exactly the **default symmetric alpha prior** case (`alpha = [0.1, ...]`), so the alpha-proposal alias can be silently corrupted in ordinary use; skewed/equal word-proposal tables can hit it too.

**Why it breaks correctness (not just mixing):** the MH acceptance ratio reads the stored `w_qw` / `prob` *intended* weights, but the samples are drawn from the *corrupted* table. The proposal density used in the Hastings ratio then does not match the distribution actually sampled, so bucket with prob 0 is never proposed and another gets double mass -> the chain no longer targets the correct posterior. Affects BOTH samplers.

**Fix:** guard the pop on both worklists, then finalize everything remaining:
```rust
while !small.is_empty() && !large.is_empty() {
    let s = small.pop().unwrap();
    let l = *large.last().unwrap();
    ...
}
for &i in small.iter().chain(large.iter()) { prob[i] = 1.0; }
```

Other alias notes: all-zero fallback correct; `gen_range(0..n)` avoids modulo bias; `Alias::build(&[])` then `sample()` panics (empty table); negative/nonfinite individual weights are not validated (a positive total does not make them valid).

## 2. [Medium] WarpLDA `set_hyper` leaves pending word-proposals drawn at the old beta

`set_hyper` updates beta (`warplda.rs:205`) and the training loop calls it right after a sweep (`python/mod.rs:2426`). At that point `prop` holds word-proposals drawn at the OLD beta (`warplda.rs:396`), but the next doc phase consumes them with acceptance algebra assuming they came from `C_wk + beta_new`. The old proposal probabilities are neither stored nor included in the ratio, so the first doc phase after every beta update has a proposal/acceptance mismatch. (The alpha alias is rebuilt correctly; the issue is the pending word-proposal buffer.) Fix: redraw pending proposals after a hyperparameter change, or store the proposal probabilities used.

## 3. [Medium] WarpLDA seeded-beta normalizer + seed hygiene

- `set_hyper` does not rebuild `seed.beta_sum_k` (`warplda.rs:157`/`190`/`205`): after `set_seeds`, `beta_sum_k = V*beta_old + |S_k|*w`; changing beta updates numerators (`beta_at`) but leaves the `V*beta_old` denominator stale. Current Python seeded-Warp does not optimize beta, so mainly an API/composability defect.
- Duplicate/OOV seeds (`warplda.rs:159`): `beta_sum_k` is incremented by `ws.len()` BEFORE invalid word IDs are discarded, and seed lists are sorted but never deduplicated. An OOV seed inflates `beta_sum_k` without adding to any `beta_at`; a duplicate inflates it multiple times while `beta_at` adds the boost once. Filter, sort, and dedup before building either structure.

## Lower-severity / API robustness
- LightLDA `n_wk` is `i32` (`lightlda.rs:117`): a word-topic cell above `i32::MAX` panics (checked) or wraps (unchecked).
- Neither sampler's public Rust constructors validate `alpha.len() == num_topics` or finite/positive entries; `set_doc_alpha` (WarpLDA) does not validate doc count / row width / positivity.
- WarpLDA `theta_into` always uses global `self.alpha`, ignoring `doc_alpha` (`warplda.rs:525`); the DMR Python path computes theta manually, so it avoids this.
- `to_topic_model` cannot preserve asymmetric seeded beta and emits a symmetric `beta_sum` (`warplda.rs:545`).

## Investigate (NOT filed as a bug): exact-posterior vs MCEM
Codex questioned whether the delayed-update (WarpLDA) and cycle-proposal (LightLDA) schemes, which draw proposals from INCLUSIVE (non-leave-one-out) counts while the acceptance uses leave-one-out target counts, EXACTLY target the collapsed-Gibbs posterior. On hand-derivation both acceptance ratios are internally consistent and match each paper's fixed-proposal formula (WarpLDA doc-phase reduces to `[(C_d,prop+alpha)(C_k,cur-1+beta_sum)] / [(C_d,cur-1+alpha)(C_k,prop+beta_sum)]`; LightLDA doc-proposal matches Eq 11 with full-count proposal ratio). Whether these are exact MCMC or a paper-sanctioned approximation is a design question, not a confirmed defect. **Settle it with a reference-free test:** enumerate all assignments of a tiny 2-topic corpus, compute the exact collapsed posterior via gamma functions, and compare long-run empirical assignment frequencies for each proposal kernel separately (detailed balance `P(s)T(s,t) = P(t)T(t,s)`). Do this AFTER fixing the alias bug (which otherwise contaminates the test).

## Suggested tests
- **Alias implied-PMF** (module-internal, inspects private fields): reconstruct `Pr(j) = (1/K) sum_i [prob_i*1(i=j) + (1-prob_i)*1(alias_i=j)]` and compare to normalized inputs for `[0.1,0.1,0.1]` (critical regression), `[0,1,0]`, `[1e-300,1,1e-300]`, all-zero, single, empty. 
- Detailed-balance / tiny-state stationary test per proposal kernel (see Investigate).
- Count invariants after every phase/token: `sum(n_k)==total_tokens`, recount-from-z == n_k, n_dk agrees with z, each `(doc,pos)` once in `word_index`, exactly one net decrement/increment per token update.
- Seed-normalization invariant: `sum_w beta_{k,w} == beta_sum_k` before/after `set_hyper`, with duplicate and OOV seeds.
- Hyperparameter-bound proposal test: change beta with pending proposals -> assert they are redrawn or evaluated at their original proposal probabilities.

_Filed from a Codex review pass; finding #1 (the `Alias::build` corruption, affecting both samplers) independently verified against the source, including the `[0.1,0.1,0.1]` floating-point trigger._

**Comment by nealcaren** [comment]
## Verification update

Working through the WarpLDA items:

**Alias::build corruption** — fixed in #431 (merged). This was the real, high-impact one (default symmetric prior).

**`set_hyper` β-update proposal mismatch — measured NEGLIGIBLE.** Structurally real: after a β step the pending word-proposals were drawn with β_old while the next doc-phase cancels the word factor assuming β_new, dropping a residual factor `R = (C_w,prop+β_new)(C_w,cur+β_old) / [(C_w,cur+β_new)(C_w,prop+β_old)]`. I instrumented a doc-phase after a β step on a planted corpus (mid-optimization, so many tokens still have `prop != cur`) and measured R and the per-token decision sensitivity:

- realistic +10% β step: mean sensitivity 0.0001, worst R = 0.999
- unrealistic 3× β jump: mean sensitivity 0.0012, worst R = 0.981
- β/3: mean 0.0004, worst R = 1.007

The large-residual case needs a proposal to a *zero-count* topic, which `q_w ∝ C_wk+β` makes rare, and β changes slowly. The effect is far below MCMC noise, so no fix (the redraw would add complexity + change the RNG stream for no measurable benefit). Same over-flag pattern as the #418 slice-sampler item.

**Seed-hygiene (duplicate/OOV seeds inflate `beta_sum_k`) — REAL, fixed in #444.** `set_seeds` added `ws.len()·seed_weight` verbatim while `beta_at` counts each distinct in-vocab seed once, breaking `Σ_w beta_at(t,w) == beta_sum_at(t)` and under-normalizing seeded φ (material when `seed_weight` is large; reachable via SeededLDA with duplicate seeds). Fix counts distinct in-vocab seeds only; regression test on the normalizer invariant.

Net: 2 of the 3 substantive items were real (alias #431, seed-hygiene #444); the β-update mismatch does not reproduce as an observable effect.

**Comment by nealcaren** [comment]
Resolved. Alias::build PMF corruption fixed in #431; WarpLDA seed-hygiene (distinct in-vocab seeds) in #444; the set_hyper pending-proposal mismatch was instrumented and measured negligible (worst realistic decision sensitivity ~1e-4, far below MCMC noise), so consciously not fixed. See the verification comment above.
